modules/mqtt/destination/mqtt-destination.c

```diff
@@ -0,0 +1,280 @@
+
+  log_template_unref(self->message);
+  self->message = message;
+}
+
+/*
+ * Utilities
+ */
+static const gchar *
+_format_stats_instance(LogThreadedDestDriver *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar stats_instance[1024];
+
+  if (((LogPipe *)d)->persist_name)
+    g_snprintf(stats_instance, sizeof(stats_instance), "%s", ((LogPipe *)d)->persist_name);
+  else
+    g_snprintf(stats_instance, sizeof(stats_instance),
+              "mqtt-destination,%s,%d,%s", self->host->str, self->port, self->topic->str);
+  return stats_instance;
+}
+
+static const gchar *
+_format_persist_name(const LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar persist_name[1024];
+
+  if (d->persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.%s", d->persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.(%s,%d,%s)", self->host->str, self->port, self->topic->str);
+
+  return persist_name;
+}
+
+static gboolean
+_check_is_valid_protocol(gchar *protocol_type)
+{
+  gchar* valid_type[] = {"tcp", "ssl", "ws", "wss"};
+  gint i;
+
+  for (i = 0; i < G_N_ELEMENTS(valid_type); ++i)
+      if (strcmp(protocol_type, valid_type[i]) == 0)
+        return TRUE;
+
+  return FALSE;
+}
+
+static void
+_set_default_value(LogPipe *s)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (self->host->len == 0)
+    g_string_assign(self->host, DEFAULT_HOST);
+
+  if (self->port == -1)
+    self->port = DEFAULT_PORT;
+
+  if (self->keepalive == -1)
+    self->keepalive = DEFAULT_KEEPALIVE;
+
+  if (self->qos == -1)
+    self->qos = DEFAULT_QOS;
+
+  if (self->protocol_type->len == 0)
+    g_string_assign(self->protocol_type, DEFAULT_PROTOCOL_TYPE);
+  
+  if (self->message == NULL)
+  {
+    self->message = log_template_new(cfg, NULL);
+    log_template_compile(self->message, DEFAULT_MESSAGE_TEMPLATE, NULL);
+  }
+
+  log_template_options_init(&self->template_options, cfg);
+}
+
+static gboolean
+_dd_init(LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  
+  _set_default_value(d);
+
+  if (!log_threaded_dest_driver_init_method(d))
+    {
+      msg_error("ERROR, while init threaded dest.");
+      return FALSE;
+    }
+
+  if (self->topic->len == 0)
+    {
+      msg_error("ERROR, topic must be set!"); 
+      return FALSE;
+    }
+  
+
+  if (self->super.batch_lines != -1)
+    {
+      msg_error("ERROR, can not be set batch_lines() in mqtt destination!");
+      return FALSE;
+    }
+
+  if (self->super.batch_timeout != -1)
+    {
+      msg_error("ERROR, can not be set batch_timeout() in mqtt destination!");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+gboolean
+_dd_deinit(LogPipe *s)
+{
+  // MQTTDestinationDriver *self = (MQTTDestinationDriver *)s;
+
+  return log_threaded_dest_driver_deinit_method(s);
+}
```

Comment:
you can drop this function in its entirety. we readd this if we need to override any functionality, for now this is purely boilerplate.